### PR TITLE
LSIF Typed: Specify the grammar for symbols

### PR DESCRIPTION
**Pull request by olafurpg**

### CODE REVIEW

File: proto/lib/codeintel/lsif_typed/lsif.proto [hunk 1]
```diff
@@ -74,34 +75,74 @@ message Document {
...
 }
 
+// Symbol is similar to a URI, it identifies a classes, methods, or local
+// variables without
+// `SymbolInformation` contains rich metadata about that symbols such as the
+// docstring.
+//
+// Symbol has a standardized string representation, which can be used
+// interchangeably with `Symbol`. The syntax for Symbol is the following:
+//   <symbol>               ::= <scheme> ' ' <package> ' ' { <descriptor> } | 'local ' <local-id>
```
Comment (olafurpg) [inline proto/lib/codeintel/lsif_typed/lsif.proto:88]: The syntax `{ a }` means `a` can be repeated

File: proto/lib/codeintel/lsif_typed/lsif.proto [hunk 2]
```diff
@@ -74,34 +75,74 @@ message Document {
...
+//   <type-parameter>       ::= '[' <name> ']'
+//   <parameter>            ::= '(' <name> ')'
+//   <name>                 ::= <identifier>
+//   <method-disambiguator> ::= <simple-identifier>
+//   <identifier>           ::= <simple-identifier> | <escaped-identifier>
+//   <simple-identifier>    ::= { <identifier-character> }
+//   <identifier-character> ::= '_' | '-' | '$' | ASCII letter or digit
+//   <escaped-identifier>   ::= '`' { <escaped-character> } '`'
+//   <escaped-characters>   ::= any UTF-8 character, escape backticks with double backtick.
+message Symbol {
```
Comment (olafurpg) [inline proto/lib/codeintel/lsif_typed/lsif.proto:110]: It should be relatively easy to implement a parser for this message. To begin with, we only need to be able to parse out the package which can be done by splitting the string by space characters as a hacky approximation.

File: proto/lib/codeintel/lsif_typed/lsif.proto [hunk 3]
```diff
@@ -74,34 +75,74 @@ message Document {
   repeated SymbolInformation symbols = 3;
 }
 
+// Symbol is similar to a URI, it identifies a classes, methods, or local
+// variables without
```
Comment (efritz) [inline proto/lib/codeintel/lsif_typed/lsif.proto:79]: Fragment

File: proto/lib/codeintel/lsif_typed/lsif.proto [hunk 4]
```diff
@@ -74,34 +75,74 @@ message Document {
...
+//   <method>               ::= <name> '(' <method-disambiguator> ').'
+//   <type-parameter>       ::= '[' <name> ']'
+//   <parameter>            ::= '(' <name> ')'
+//   <name>                 ::= <identifier>
+//   <method-disambiguator> ::= <simple-identifier>
+//   <identifier>           ::= <simple-identifier> | <escaped-identifier>
+//   <simple-identifier>    ::= { <identifier-character> }
+//   <identifier-character> ::= '_' | '-' | '$' | ASCII letter or digit
+//   <escaped-identifier>   ::= '`' { <escaped-character> } '`'
+//   <escaped-characters>   ::= any UTF-8 character, escape backticks with double backtick.
```
Comment (efritz) [inline proto/lib/codeintel/lsif_typed/lsif.proto:105]: We should figure out how we'd want to display this in a hierarchical symbol sidebar to make sure we get the correct set of common things across languages.

Comment (olafurpg) [reply proto/lib/codeintel/lsif_typed/lsif.proto:108]: We can introduce the concept of an "owner" of a symbol, which the symbol you get by removing the last descriptor.

File: proto/lib/codeintel/lsif_typed/lsif.proto [hunk 5]
```diff
@@ -25,7 +25,8 @@ message Index {
   Metadata metadata = 1;
   // Documents that belong to this index.
   repeated Document document = 2;
-  // Symbols that are referenced from this index and not defined in this index.
+  // (optional) Symbols that are referenced from this index and not defined in this index.
+  // Only used to provide documentation for symbols that are
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply proto/lib/codeintel/lsif_typed/lsif.proto]: Fixed

**Review by efritz — APPROVED** [review]

**Pull request merged**